Debugging notebooks

seeM opened the pull request:
This PR adds support for debugging notebooks, addressing #3906.

This is a pretty big change so I expect there to be rough edges. Where possible I'd like to defer fixes to be in separate PRs so we can get the foundation in.

### Release Notes

#### New Features

- You can now debug Python Jupyter notebooks.

#### Bug Fixes

- N/A

### QA Notes

* Open a Python notebook and try the Debug Cell command.
* Try out edge cases like interrupting a cell while debugging, restarting the runtime, debugging function calls across cells, etc.

seeM committed:
wip: debug notebook cells

refactoring + a few fixes

seeM committed:
extract positron-runtime-debug extension

seeM committed:
debug the cell if provided, otherwise currently selected

seeM committed:
fix breakpoint in function defined in another cell

seeM committed:
wip: client-side cell source mapping

seeM committed:
cleaning up

seeM committed:
rename

seeM committed:
disposable store

seeM committed:
wip: refactoring

seeM committed:
runtime-specific output channel

seeM committed:
clean up runtime debugger output channel name

seeM committed:
reject on error

seeM committed:
more cleanup

seeM committed:
fix output channel lifecycle

seeM committed:
rename

seeM committed:
rename again

seeM committed:
forgot to commit these

seeM committed:
fix bug when debugging first cell

seeM committed:
inline

seeM committed:
unused imports

seeM committed:
working post refactor

seeM committed:
extract source mapper

seeM committed:
more shuffling

seeM committed:
cleanup dap transformer

seeM committed:
finalizing debug protocol transformer interface

seeM committed:
not using rpc anymore - separate request/reply handlers for now

seeM committed:
refactor transformer

seeM committed:
it works!

seeM committed:
transform both ways

seeM committed:
cleaning up unused code paths

seeM committed:
naming

seeM committed:
refactor dependency directon

seeM committed:
cleaning up source map

seeM committed:
extract source mapper

seeM committed:
finalizing source map abstractions

seeM committed:
cleaning up

seeM committed:
renames

seeM committed:
docs

seeM committed:
`Disposable` helper

seeM committed:
take `cell` instead of notebook and cell index

seeM committed:
naming

seeM committed:
rename again

seeM committed:
more renames

seeM committed:
todo

seeM committed:
remove unusued

seeM committed:
only refresh if initialized

seeM committed:
restore state just before final initialized event

seeM committed:
throw instead of silently returning

seeM committed:
add `LanguageRuntimeSession.runtimeInfo`

seeM committed:
declare debugger support

seeM committed:
finalize runtimes declaring debugging support

seeM committed:
refine positron runtime debug api

seeM committed:
fix types

seeM committed:
remove debugprotocol dep

seeM committed:
comments

jmcphers reviewed (dismissed):
Only looked in detail at the API and supervisor bits.

austin3dickey reviewed (commented):
Maybe I'm missing something but I can't quite figure out how to test this? Here's what I tried:
- open a python notebook as a Jupyter (not Positron) notebook
- set a breakpoint
- the run cell button doesn't stop at the breakpoint
- if I click the dropdown next to the run button, sometimes Debug Cell is greyed out and I can't find the _Notebook: Debug Cell_ command
- If I can press the button or run the command, it prompts me to create a new launch.json file
- there's no option for notebook debugging when creating a launch.json configuration

I do think I've got this extension running because `positron.positron-runtime-debugger` is in the list of running extensions.

seeM committed:
surface runtime errors during initialization

seeM committed:
logging

seeM committed:
fix notebook interpreter supports debugging context

seeM committed:
more generic context

seeM committed:
fix enablement clause

seeM committed:
make `debug` async and return the rpc response

seeM committed:
simplify `debug`

seeM committed:
Revert "more generic context"

This reverts commit 787f2c76e0b3d9a74fba7a3a4c5435cde7b5daff.

seeM committed:
Revert "fix enablement clause"

This reverts commit 577bfd9a8299b38c12d9927cea217cd2f27799a0.

seeM committed:
fix bug; rewrite tests

seeM committed:
pyright

seeM committed:
test

seeM commented:
@austin3dickey and I paired and discovered a few issues. I've updated the branch so that more helpful runtime errors are shown to the user. For example, if you have the `python.languageServerDebug` setting enabled, you will now see a more informative error and can view the full traceback:

<img width="268" height="284" alt="image" src="https://github.com/user-attachments/assets/9de585cd-99a8-4261-a534-a07d303f0f20" />

I've also updated the `notebookInterpreterSupportsDebugging` context so that the Debug Cell command is only greyed out when it should be + added some unit tests.

Austin could you please give it another try?

seeM committed:
rename cell controller; remove refresh state and depend directly on path encoder

seeM committed:
stop debugging listeners

seeM committed:
disable debug cell for debugged notebooks

seeM committed:
disable updates on edits, for now

seeM committed:
comments

seeM committed:
oops

seeM committed:
localize

nstrayer reviewed (dismissed):
This is a crazy amount of (impressive) work. 

I tested it in a variety of scenarios and couldn't break it (with an exception noted in the inline comments.)

I added just a few comments about things but they're all nits. 

Thank you for this!

austin3dickey reviewed (dismissed):
Thank you for all the hard work!! It works very well for me.

~I'm still seeing that "Debug Cell" is greyed out at first when I first open notebooks. Maybe it's because the session is starting? Once I tab to a different editor and come back, it's enabled again. Happy to defer this to a new issue.~ edit: never mind, this was a local build issue.

Congrats! I love the error popup!

seeM commented on extensions/positron-runtime-debugger/src/notebookLocationMapper.ts:
```diff
@@ -0,0 +1,136 @@
+			this.refresh();
+		}));
+
+		// TODO: Do we need to update mappings when the notebook changes?
+		//       Probably need to throttle this if so.
```
Thanks for testing! I'll remove the comment

seeM commented on extensions/positron-runtime-debugger/src/notebookDebugAdapterFactory.ts:
```diff
@@ -0,0 +1,80 @@
+		if (outputChannel) {
+			return outputChannel;
+		}
+		const newOutputChannel = this._register(createDebuggerOutputChannel(runtimeSession));
+		this._outputChannelByRuntimeSessionId.set(runtimeSession.metadata.sessionId, newOutputChannel);
```
Hmm I can't think of any. If we dispose the channel, it seems to remove it from the Outputs pane

seeM committed:
address reviews
